geopandas/geodataframe.py

```diff
@@ -589,17 +669,27 @@ def __getitem__(self, key):
         geo_col = self._geometry_column_name
         if isinstance(result, Series) and isinstance(result.dtype, GeometryDtype):
             result.__class__ = GeoSeries
-            result.crs = self.crs
             result._invalidate_sindex()
         elif isinstance(result, DataFrame) and geo_col in result:
             result.__class__ = GeoDataFrame
-            result.crs = self.crs
             result._geometry_column_name = geo_col
             result._invalidate_sindex()
         elif isinstance(result, DataFrame) and geo_col not in result:
             result.__class__ = DataFrame
         return result
 
+    def __setitem__(self, key, value):
+        """
+        Overwritten to preserve CRS of GeometryArray in cases like
+        df['geometry'] = [geom... for geom in df.geometry]
+        """
+        if key == self._geometry_column_name:
```

Comment:
This comparison is going to fail if `key` is array like. 
Eg consider:

```
df = geopandas.read_file(geopandas.datasets.get_path('naturalearth_lowres'))  
df[np.array(['pop_est', 'continent'])] = 1
# or (although this doesn't make much sense logically, but key can be a boolean series/array)
df[df['continent'] == 'Africa'] = 1
```

so we will need to add a check like `if not pd.api.types.is_list_like(key) and key == ..`